docs/src/developer/design/user_api/add_target.rst

```diff
@@ -0,0 +1,324 @@
+   API ultimately map to underlying objects. a) Functions related to creating
+   an executable. b) Functions related to creating a library. c) Conditional
+   functions wrapping functions from a) and b).
+
+:numref:`fig_add_target_functions` summarizes the functions implementing the
+front end of CMaize's target component. As part of the user :term:`API`, the
+front end is function-based (see :ref:`functional_style`). Each box in
+:numref:`fig_add_target_functions` represents a function ("snake_case" labels),
+section of a function (boxes labeled with phrases), or an object
+("UpperCamelCase" labels). Nested boxes represent the parts of the function.
+
+As mentioned above, one of the main considerations for CMaize's add target
+functions is :ref:`at_recording_targets`. In :numref:`fig_add_target_functions`
+recording the target information is done in the boxes labeled "Notify Project".
+The other major consideration for CMaize's add target functions is
+:ref:`at_succinctness`. Succinctness is the motivation for the dispatch steps.
+To clarify, CMaize is able to avoid requiring the user to specify some of the
+target's properties (e.g., language, header-only) simply by analyzing the
+:term:`source files <source file>` the user provides. Conceivably it is possible
+to glean even more details (for example the C++/Python version used) from the
+source file contents; however, CMaize currently makes no attempt to do so.
+
+While the API of the add target functions is language agnostic, consideration
+:ref:`at_coding_languages` means the backend can not be. As shown in
+:numref:`fig_add_target_functions`, both ``cmaize_add_executable`` and
+``cmaize_add_library`` ultimately dispatch to language-specific target objects.
+In this sense, CMaize's add target functions can be thought of as factory
+routines.
+
+***************
+Add Target APIs
+***************
+
+The APIs for ``cmaize_add_executable`` and ``cmaize_add_library`` were
+briefly introduced in the :ref:`designing_cmaizes_user_api` section. Generally
+speaking the two functions work similarly aside from the fact that
+``cmaize_add_executable``/\ ``cmaize_add_library`` will ultimately map to
+CMake's ``add_executable``/\ ``add_library`` function, respectively (the calls
+to the CMake functions happen inside the target objects). Given the similar APIs
+we presently limit our focus to ``cmaize_add_library``:
+
+.. code-block:: CMake
+
+   # A typical C++ invocation
+   cmaize_add_library(
+       "<target name>"
+       SOURCE_DIR "<directory containing private source files>"
+       INCLUDE_DIRS "<directory(s) containing public source files>"
+       DEPENDS "<name of dependency 0>" "<name of dependency 1>"
+   )
+
+   # A typical Python invocation
+   cmaize_add_library(
+       "<target name>"
+       INCLUDE_DIRS "<directory(s) containing the Python library>"
+       DEPENDS "<name of dependency 0>" "<name of dependency 1>"
+   )
```

Comment:
@zachcran what do you think of this API? I'm wondering if `INCLUDE_DIRS` for Python should map to `PYTHONPATH` and `SOURCE_DIR` should be the contents of the library?

Reply:
I like `SOURCE_DIR` pointing to the library code. I'm not so sure about `INCLUDE_DIRS` mapping to `PYTHONPATH`. It could definitely work, but I think you usually want to set up a virtual environment and install everything there. It seems like making it this accessible would encourage users away from the best practices of venvs and installable packages. Maybe it could be pushed to an optional `PYTHONPATH` parameter?

Reply:
We can punt on `PYTHONPATH` for now. When managing hybrid language projects its not so easy as "install everything in a virtual environment", e.g., if I'm trying to test the Python bindings for the C++ library I just built it's going to be sitting in the build directory, not installed in the virtual environment.

Reply:
For testing things like the Python bindings you mentioned, could we add something like `PYTHONPATH=$PYTHONPATH:/path/to/build/directory <actual python command>` when calling `add_test()` when CMaize creates Python tests.